feat: add possibility to disable default redirect for prefix_and_default strategy

Renhor opened the pull request:
Regarding to issue #1401 

Added property to config: 
```
i18n: {
  disableDefaultRedirect: boolean | string[] // false by default
}
```
It works only for strategy **prefix_and_default**.
It disables redirect for routes are prefixed with default language, or we can choose custom path pattern for disabling redirect behaviour:
```
i18n: {
  disableDefaultRedirect: [
    '/about',
    '/blog/*'
  ]
}
```

**UPD:**

Added property to config: 
``` 
type PrefixAndDefaultRule = 'default' | 'prefix'

i18n: {
  prefixAndDefaultRule?: PrefixAndDefaultRule
}
```

By setting the value **prefixAndDefaultRule = 'prefix'**  we are no longer redirected from prefixed routes and if there is no prefix  then the other routers will also have no prefix.

For now, I decided not to implement an array with a list of routes for which the default behavior is disabled

Commit:
feat: add possibility to disable default redirect for prefix_and_default strategy

rchl commented on src/templates/plugin.utils.js:
```diff
@@ -216,6 +216,40 @@ export function mergeAdditionalMessages (i18n, additionalMessages, localeCodes,
+
+  return !disableDefaultRedirect
+    .map((patternString) => {
+      const isPattern = patternString.includes('*')
```
Looks like `*` is only supported at the end so you could use `endsWith` here.

Renhor replied:
Currently i abandon the route array cuz i can't find it useful in the current implementation

rchl commented on src/templates/plugin.main.js:
```diff
@@ -153,14 +153,20 @@ export default async (context) => {
     let redirectPath = ''
 
     const isStaticGenerate = process.static && process.server
+    const isSameLocale = getLocaleFromRoute(route) !== newLocale
```
The variable name is wrong. Should be `isDifferentLocale`.

rchl commented on src/templates/utils-common.js:
```diff
@@ -223,3 +223,22 @@ export function setLocaleCookie (locale, res, { useCookie, cookieDomain, cookieK
+ * @param  {readonly string[]} localeCodes
+ * @return {string}
+ */
+export function clearPathFromLocale (pathString, localeCodes) {
```
Naming feels off. Maybe
```suggestion
export function removeLocaleFromPath (pathString, localeCodes) {
```

rchl commented on src/templates/utils-common.js:
```diff
@@ -223,3 +223,22 @@ export function setLocaleCookie (locale, res, { useCookie, cookieDomain, cookieK
+ * @return {string}
+ */
+export function addSlashToPath (pathString) {
+  return pathString.endsWith('/') ? pathString : `${pathString}/`
```
This seems a bit too naive and will likely fail if there are query params in the path (`route.fullPath`). This module uses `ufo` which should have more robust implementation of that (`withTrailingSlash` maybe).

Renhor replied:
Thanks for advice, i begin to use ufo

rchl commented on types/index.d.ts:
```diff
@@ -52,9 +52,12 @@ export interface BaseOptions {
 export interface Options extends BaseOptions {
   baseUrl?: string | ((context: NuxtContext) => string)
   detectBrowserLanguage?: DetectBrowserLanguageOptions | false
+  disableDefaultRedirect?: DisableDefaultRedirect;
```
The naming of that option is a bit too generic for my taste since it applies only in very specific case, only when using specific strategy. We can think more about it after other changes are addressed.

Renhor replied:
What about 
```
type PrefixAndDefaultRule = 'default' | 'prefix'

interface PrefixAndDefaultRules {
  switchLocal: PrefixAndDefaultRule // 'default'
  routing: PrefixAndDefaultRule // 'default'
}

i18n: {
  prefixAndDefaultRules?: PrefixAndDefaultRules
}
```

rchl commented on src/templates/plugin.main.js:
```diff
@@ -153,14 +153,20 @@ export default async (context) => {
-      (getLocaleFromRoute(route) !== newLocale || (options.strategy === Constants.STRATEGIES.PREFIX_AND_DEFAULT && newLocale === options.defaultLocale))
+      // locale, in which case we might still redirect as we prefer unprefixed route in this case, but let user a
+      // possibility to disable this behavior by switching disableDefaultRedirect option to true.
+      (isSameLocale || (isPrefixAndDefaultStrategy && isDefaultLocale && isEnabledRedirectForRoute))
```
I feel like having this would only solve part of the issue as linking to such page using `localePath('about')` would still resolve and navigate to locale-less path. If you are willing to work on it more to also handle that case then I'll be willing to accept it.

Note that if this would be handled in the path resolving code (`localePath` and co.) then you probably wouldn't even need to change this part.

rchl reviewed (changes requested):
This would also need documentation to be updated with new option and some tests.

Commit:
fix: better namings, replace custom trailing slash function by ufo withTrailingSlash

Commit:
feat: add playground for disableRedirect

Commit:
feat: routing functions now resolves prefixed routes if option(prefixAndDefaultRedirect) enabled

Commit:
fix: update playground for prefixAndDefaultRedirect playground

Commit:
feat: refactor feature, remove perPage check for prefix

Renhor commented on src/templates/plugin.main.js:
```diff
@@ -153,18 +153,22 @@ export default async (context) => {
     ) {
-      // The current route could be 404 in which case attempt to find matching route using the full path since
-      // "switchLocalePath" can only find routes if the current route exists.
-      const routePath = app.switchLocalePath(newLocale) || app.localePath(route.fullPath, newLocale)
```
[rchl](https://github.com/rchl) With my current implementation switchLocalePath() always redirects me to the prefixed path and I couldn't think of anything better yet :(

Renhor commented:
[rchl](https://github.com/rchl) Thanks for the quick review. Unfortunately, I will only be able to continue working on the task today. I decided to abandon the route array and changed the implementation following your advice. Updated the information in the first post

Commit:
fix: fix typos

Commit:
feat: update docs, add references for prefixAndDefaultRules option

rchl commented on docs/content/en/api.md:
```diff
@@ -28,12 +28,18 @@ All [Vue I18n properties and methods](http://kazupon.github.io/vue-i18n/api/#vue
 
   - **Arguments**:
     - locale: (type: `string`)
+    - forcePrefix: (type: `boolean`)
```
I'd want to keep the changes as simple as possible while fixing your use case.

Is this new argument necessary? I have hard time imagining why would only sometimes want or not want the locale prefix.

Renhor replied:
By adding a split for `{ routing: 'default', switchLocale: 'default' }`, I thought that in some situations we wouldn't want to globally change the behavior of the method to something else, so I decided to add this argument.

However, I now think that such a situation is actually hard to imagine.

rchl commented on docs/content/en/options-reference.md:
```diff
@@ -186,6 +186,18 @@ Set this to `true` when using different domains for each locale. If enabled, no
+## `prefixAndDefaultRules`
+
+- type: `object`
+- default: `{ routing: 'default', switchLocale: 'default' }`
```
Is it necessary to have separate settings for `routing` and `switchLocale`?

Renhor replied:
I agree, let's keep a single property for the behavior

rchl reviewed (changes requested):
I have not looked properly at the changes yet because many tests are failing.

Commit:
fix: fixes after code review

Renhor commented:
Corrected the code that was causing the tests to crash

juliomontilla100 commented:
I need this feature ! please, thanks... @Renhor any update bro?

DanielHefti commented:
This feature would be highly appreciated. Get in touch with me, if there is anything I can do.
